src/theory/quantifiers/quantifiers_rewriter.cpp

```diff
@@ -1743,6 +1744,90 @@ Node QuantifiersRewriter::computeVarElimination(Node body,
     }
     Trace("var-elim-quant") << "Return " << body << std::endl;
   }
+  // Leibniz equality elimination
+  if (d_opts.quantifiers.leibnizEqElim)
+  {
+    if (body.getKind() == Kind::OR && body.getNumChildren() == 2) // the body must have exactly 2 children
+    {
+      Node termA = body[0];
+      Node termB = body[1];
+
+      // helper
+      auto matchApply = [&](Node lit, Node &op, std::vector<Node> &argsOut, bool &neg) -> bool {
```

Comment:
Can you make this its own utility method?
Also, the two cases can be merged to one i.e. start with:
```
neg = lit.getKind()==Kind::NOT;
Node atom = neg ? lit[0] : lit;
... check atom instead of lit
```